File: src/main/java/nuclearcoder/discordbot/command/custom/CmdCustomCommand.java
package nuclearcoder.discordbot.command.custom;

import nuclearcoder.discordbot.BotUtil;
import nuclearcoder.discordbot.NuclearBot;
import nuclearcoder.discordbot.command.Command;
import nuclearcoder.util.ArgumentFormatter;
import sx.blah.discord.handle.obj.IMessage;
import sx.blah.discord.handle.obj.IUser;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class CmdCustomCommand implements Command {

    // used when there is only one format
    private String singleGuildID;
    private ArgumentFormatter singleFormat;

    // used when there are 2+ formats
    private Map<String, ArgumentFormatter> formats;

    public CmdCustomCommand(String guildID, String format)
    {
        this.singleGuildID = guildID;
        this.singleFormat = buildFormatter(format);

        this.formats = null;
    }

    private ArgumentFormatter buildFormatter(String format)
    {
        return new ArgumentFormatter(format);
    }

    public void put(String guildID, String format)
    {
        if (formats == null)
        {
            formats = new HashMap<>();
            formats.put(singleGuildID, singleFormat);

            singleGuildID = null;
            singleFormat = null;
        }

        formats.put(guildID, buildFormatter(format));
    }

    // returns true if empty after removal
    public boolean remove(String guildID)
    {
        boolean isEmpty = (formats == null);

        if (isEmpty)
        {
            singleGuildID = null;
            singleFormat = null;
        }
        else
        {
            formats.remove(guildID);
            if (formats.isEmpty())
            {
                formats = null;
            }
        }

        return isEmpty;
    }

    private ArgumentFormatter getFormatter(String guildID)
    {
        if (formats == null)
        {
            return guildID.equals(singleGuildID) ? singleFormat : null;
        }
        else
        {
            return formats.get(guildID);
        }
    }

    @Override public void execute(NuclearBot bot, IMessage message, String command, String[] args)
    {
        String guildID = message.getGuild().getID();
        ArgumentFormatter formatter = getFormatter(guildID);

        if (formatter != null)
        {
            IUser author = message.getAuthor();

            String[] cmdArgs = Arrays.copyOfRange(args, 1, args.length);
            String formattedMessage = formatter.format(author.mention(), cmdArgs);
            if (formattedMessage != null)
            {
                BotUtil.sendMessage(message.getChannel(), formattedMessage);
            }
        }
    }

}

File: src/main/java/nukebot/command/custom/CmdCustomCommand.kt
package nukebot.command.custom

import nukebot.NuclearBot
import nukebot.command.Command
import nukebot.util.rSendMessage
import nukeutils.ArgumentFormatter
import sx.blah.discord.handle.obj.IMessage
import java.util.*

class CmdCustomCommand(guildID: String, format: String) : Command {

    // used when there is 1 format
    private var singleGuildID = guildID
    private var singleFormat = ArgumentFormatter(format)

    // used when there are 2+ formats
    private var formats: MutableMap<String, ArgumentFormatter>? = null

    fun put(guildID: String, format: String) {
        if (formats == null) {
            formats = HashMap()
            formats?.put(singleGuildID, singleFormat)
        }
        formats?.put(guildID, ArgumentFormatter(format))
    }

    // returns true if empty after removal
    fun remove(guildID: String): Boolean {
        if (formats == null) return true

        formats?.remove(guildID)
        if (formats?.isEmpty() ?: true) { // if formats is empty, free it
            formats = null
        }

        return false
    }

    private fun getFormatter(guildID: String): ArgumentFormatter? {
        return if (formats == null) {
            if (guildID == singleGuildID) singleFormat else null
        } else {
            formats?.get(guildID)
        }
    }

    override fun execute(bot: NuclearBot, message: IMessage, command: String, args: Array<String>) {
        val formatter = getFormatter(message.guild.id)

        if (formatter != null) {
            val author = message.author

            val formattedMessage = formatter.format(author.mention(), *args.sliceArray(1..args.lastIndex))
            if (formattedMessage != null) {
                message.channel.rSendMessage(formattedMessage)
            }
        }
    }

}
